components/service/sync-autofill/src/main/java/mozilla/components/service/sync/autofill/DefaultCreditCardValidationDelegate.kt

```diff
@@ -0,0 +1,69 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.service.sync.autofill
+
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import mozilla.components.concept.storage.CreditCard
+import mozilla.components.concept.storage.CreditCardEntry
+import mozilla.components.concept.storage.CreditCardNumber
+import mozilla.components.concept.storage.CreditCardValidationDelegate
+import mozilla.components.concept.storage.CreditCardValidationDelegate.Result
+import mozilla.components.concept.storage.CreditCardsAddressesStorage
+
+/**
+ * A delegate that will check against the [CreditCardsAddressesStorage] to determine if a given
+ * [CreditCard] can be persisted and returns information about why it can or cannot.
+ *
+ * @param storage An instance of [CreditCardsAddressesStorage].
+ */
+class DefaultCreditCardValidationDelegate(
+    private val storage: Lazy<CreditCardsAddressesStorage>
+) : CreditCardValidationDelegate {
+
+    private val coroutineContext by lazy { Dispatchers.IO }
+
+    override suspend fun validate(creditCard: CreditCardEntry): Result =
+        withContext(coroutineContext) {
+            val creditCards = storage.value.getAllCreditCards()
+
+            val foundCreditCard = if (creditCards.isEmpty()) {
+                // No credit cards exist in the storage, create a new credit card to the storage.
+                null
+            } else {
+                // Found a matching guid and credit card number -> update
+                creditCards.find {
+                    it.guid == creditCard.guid &&
+                        decrypt(it.encryptedCardNumber)?.number == creditCard.number
+                }
+                    // Found a matching guid -> update
+                    ?: creditCards.find { it.guid == creditCard.guid }
```

Comment:
In what cases are we going to have a `creditCard.guid` here? See my question above about how we get here - perhaps we can remove this check entirely?

(although, it looks useful in case we somehow _do_ have a `guid`... i wish we had a CreditCard type without that field entirely.)

Reply:
User that autofilled a credit card and then change any of the cc number, year, or month will have a `guid` associated with the autofilled credit card.